src/librustc_yk_sections/mir_cfg.rs

```diff
@@ -18,20 +16,95 @@ use rustc::ty::TyCtxt;
 
 use rustc::hir::def_id::DefId;
 use rustc::mir::{
-    Mir, TerminatorKind, Operand, Constant, StatementKind, BasicBlock, BasicBlockData, Terminator
+    Mir, TerminatorKind, Operand, Constant, StatementKind, BasicBlock, BasicBlockData, Terminator,
+    Place, Rvalue, Statement, Local, PlaceBase
 };
 use rustc::ty::{TyS, TyKind, Const, LazyConst};
 use rustc::util::nodemap::DefIdSet;
 use std::path::PathBuf;
 use std::fs::File;
+use std::convert::TryFrom;
 use rustc_yk_link::YkExtraLinkObject;
 use std::fs;
 use std::error::Error;
+use std::cell::{Cell, RefCell};
+use rustc_data_structures::indexed_vec::{IndexVec, Idx};
+use rustc_data_structures::graph::dominators::Dominators;
 use ykpack;
 
 const SECTION_NAME: &'static str = ".yk_tir";
 const TMP_EXT: &'static str = ".yk_tir.tmp";
 
+type TirVarIndex = u32;
+
+/// A conversion context holds the state needed to perform the conversion to (pre-SSA) TIR.
+struct ConvCx<'a, 'tcx, 'gcx> {
+    /// The compiler's god struct. Needed for queries etc.
+    tcx: &'a TyCtxt<'a, 'tcx, 'gcx>,
+    /// The definition sites of TIR variable in terms of basic blocks.
+    def_sites: RefCell<Vec<Vec<BasicBlock>>>,
+    /// The next TIR variable index to be issued.
+    next_tir_var: Cell<TirVarIndex>,
+    /// A mapping from MIR variables to TIR variables.
+    var_map: RefCell<IndexVec<Local, Option<TirVarIndex>>>,
+}
+
+impl<'a, 'tcx, 'gcx> ConvCx<'a, 'tcx, 'gcx> {
+    fn new(tcx: &'a TyCtxt<'a, 'tcx, 'gcx>, _mir: &Mir<'tcx>) -> Self {
+        let var_map = IndexVec::new();
+
+        Self {
+            tcx,
+            def_sites: RefCell::new(Vec::new()),
+            next_tir_var: Cell::new(0),
+            var_map: RefCell::new(var_map),
+        }
+    }
+
+    /// Issues a fresh TIR variable index.
+    fn next_tir_var(&self) -> TirVarIndex {
+        let var_idx = self.next_tir_var.get();
+        self.next_tir_var.set(var_idx + 1);
+        var_idx
+    }
+
+    /// Get the TIR variable corresponding with the MIR variable, creating a fresh one if needed.
```

Comment:
Should this function be called `mir_var_to_tir` or similar?

Reply:
It could be... I don't think it improves the meaning much though!

"give me an index for this MIR variable, make a new index if we havne't seen it yet".

Reply:
In general, Rust has a dislike of `get` functions, and I've come to think this is generally a good idea.

Reply:
how about just `tir_var()` then?